gateway/src/services/config-manager-v2.ts

```diff
@@ -267,10 +323,31 @@ export class ConfigManagerV2 {
   loadConfigRoot(configRootPath: string) {
     // Load the config root file.
     const configRootFullPath: string = fs.realpathSync(configRootPath);
+    const configRootTemplateFullPath: string = path.join(
+      ConfigTemplatesDir,
+      'root.yml'
+    );
     const configRootDir: string = path.dirname(configRootFullPath);
     const configRoot: ConfigurationRoot = yaml.load(
       fs.readFileSync(configRootFullPath, 'utf8')
     ) as ConfigurationRoot;
+    const configRootTemplate: ConfigurationRoot = yaml.load(
+      fs.readFileSync(configRootTemplateFullPath, 'utf8')
+    ) as ConfigurationRoot;
+
+    // version control to only handle upgrades
+    if (configRootTemplate.version > configRoot.version) {
+      // run migration in order if available
+      for (
```

Comment:
Have you thought about the case where the current config is at version 10, the newest version is version 13 - and you've got migration functions defined for versions 1, 3, 6, 10, 12, 13? Simply running everything from version 1 doesn't work.

Reply:
Also, if you're executed any migrations, you should save the configurations after `addNamespace()`.